### PR TITLE
fix: dialling duplicated addresses

### LINKED ISSUE #1409: Duplicated addresses in _createDialTarget
- **Version**: 0.39.2 https://github.com/ChainSafe/lodestar/blob/ca27458b59c66c9d3da7dd95322e479f78407269/packages/beacon-node/package.json#L144

- **Platform**: Linux

- Severity: High

#### Description:
Running local devnets with Lodestar nodes couldn't dial with each other due to this error
```
Oct-05 22:08:39.721[rest]            error: Req req-1 gossipsubAttack error  dial with more addresses than allowed 5 > 4
Error: dial with more addresses than allowed 5 > 4
    at DefaultDialer._createDialTarget (file:///home/lion/Code/eth2.0/lodestar-attack/node_modules/libp2p/src/connection-manager/dialer/index.ts:264:21)
    at DefaultDialer._createCancellableDialTarget (file:///home/lion/Code/eth2.0/lodestar-attack/node_modules/libp2p/src/connection-manager/dialer/index.ts:217:26)
    at DefaultDialer.dial (file:///home/lion/Code/eth2.0/lodestar-attack/node_modules/libp2p/src/connection-manager/dialer/index.ts:179:24)
    at EventTarget.openConnection (file:///home/lion/Code/eth2.0/lodestar-attack/node_modules/libp2p/src/connection-manager/index.ts:521:26)
    at EventTarget.dial (file:///home/lion/Code/eth2.0/lodestar-attack/node_modules/libp2p/src/libp2p.ts:382:12)
```

I manually called dial submitting a single Multiaddr with `/ip4/127.0.0.1/tcp/9000/p2p/16Uiu2HAm6fBRJ3MWKUpVhscGdPkrUJjB943JfG8hbg41k9soaNpw`. But how can there be 5 addresses if I only ever provided a single IP port combo?

Then added a console log to print the multiaddr:
```
Trace: [
  <Multiaddr 047f000001062328a5032700250802122102a6e664edcd4b990208e101fe7486282a205d4a88edae4be6d54ead4257bb1f98 - /ip4/127.0.0.1/tcp/9000/p2p/16Uiu2HAm6fBRJ3MWKUpVhscGdPkrUJjB943JfG8hbg41k9soaNpw>,
  <Multiaddr 047f000001062328a5032700250802122102a6e664edcd4b990208e101fe7486282a205d4a88edae4be6d54ead4257bb1f98 - /ip4/127.0.0.1/tcp/9000/p2p/16Uiu2HAm6fBRJ3MWKUpVhscGdPkrUJjB943JfG8hbg41k9soaNpw>,
  <Multiaddr 04c0a80214062328a5032700250802122102a6e664edcd4b990208e101fe7486282a205d4a88edae4be6d54ead4257bb1f98 - /ip4/192.168.2.20/tcp/9000/p2p/16Uiu2HAm6fBRJ3MWKUpVhscGdPkrUJjB943JfG8hbg41k9soaNpw>,
  <Multiaddr 04c0a802ef062328a5032700250802122102a6e664edcd4b990208e101fe7486282a205d4a88edae4be6d54ead4257bb1f98 - /ip4/192.168.2.239/tcp/9000/p2p/16Uiu2HAm6fBRJ3MWKUpVhscGdPkrUJjB943JfG8hbg41k9soaNpw>,
  <Multiaddr 04c0a802ef062328a5032700250802122102a6e664edcd4b990208e101fe7486282a205d4a88edae4be6d54ead4257bb1f98 - /ip4/192.168.2.239/tcp/9000/p2p/16Uiu2HAm6fBRJ3MWKUpVhscGdPkrUJjB943JfG8hbg41k9soaNpw>
]
    at DefaultDialer._createDialTarget (file:///home/lion/Code/eth2.0/lodestar-attack/node_modules/libp2p/src/connection-manager/dialer/index.ts:264:13)
    at DefaultDialer._createCancellableDialTarget (file:///home/lion/Code/eth2.0/lodestar-attack/node_modules/libp2p/src/connection-manager/dialer/index.ts:217:26)
    at DefaultDialer.dial (file:///home/lion/Code/eth2.0/lodestar-attack/node_modules/libp2p/src/connection-manager/dialer/index.ts:179:24)
    at EventTarget.openConnection (file:///home/lion/Code/eth2.0/lodestar-attack/node_modules/libp2p/src/connection-manager/index.ts:521:26)
    at EventTarget.dial (file:///home/lion/Code/eth2.0/lodestar-attack/node_modules/libp2p/src/libp2p.ts:382:12)
```

so:
- Why are addresses duplicated?
- Why are IPs `192.168.2.20` and `192.168.2.239` added to the targets?

#### Steps to reproduce the error:
On a local devnet
- Start node 1 with dev command
- Start node 2 with dev command
- Call connectPeer API on node 2 with node 1 identity

### PR DESCRIPTION
**Pull request by mpetrunic**
resolves #1409

### CODE REVIEW

File: test/dialing/direct.node.ts [hunk 1]
```diff
@@ -103,6 +106,19 @@ describe('Dialing (direct, TCP)', () => {
     await connection.close()
   })
 
+  it('should be able to connect to remote node with duplicated addresses', async () => {
```
Comment (achingbrain) [inline test/dialing/direct.node.ts:109]: Please can an assertion be added here to ensure there are multiple addresses in play?

**Review by achingbrain — CHANGES_REQUESTED** [review]

File: test/dialing/direct.node.ts (hunk 1 above)
Comment (mpetrunic) [reply test/dialing/direct.node.ts:109]: Not sure I understand. I've set `maxAddrsToDial` to 1 so without filtering duplicates, this test fails as it tries to dial with 2 duplicated addresses. 

If you meant to assert there were duplicates before filtering, I don't think it's possible without code instrumentation as it happens inside the pipe: https://github.com/libp2p/js-libp2p/blob/master/src/connection-manager/dialer/index.ts#L238 (one address comes from dial param and other from resolving address in peer book).

This bug doesn't happen if address book contains resolved duplicate^^

**Review by achingbrain — APPROVED** [review]

File: src/connection-manager/dialer/index.ts [hunk 2]
```diff
@@ -259,6 +259,8 @@ export class DefaultDialer implements Startable, Dialer {
       async (source) => await all(source)
     )
 
+    addrs = [...new Set(addrs)]
```
Comment (dapplion) [inline src/connection-manager/dialer/index.ts:262]: No concerns for performance and unnecessary allocation of tons of memory for just selecting a list of addresses?

Comment (mpetrunic) [reply src/connection-manager/dialer/index.ts:262]: Don't think performance or memory will be issue here as we are mostly talking about <10addresses

Comment (dapplion) [reply src/connection-manager/dialer/index.ts:262]: It's not that it's an issue, it that it is unnecessary to tackle this problem this way. This entire function contains multiple un-explicable design choices with the common theme of zero regard for performance in exchange for no clear benefit. A non-piped version would be even easier to read and mantain, and would not require to loop **again** in the Set iteration to drop duplicates

Comment (dapplion) [reply src/connection-manager/dialer/index.ts:262]: As an extreme example: this solution below is correct, and won't affect performance in libp2p overall. But we should never accept it regardless
```suggestion
    addrs = [...new Set([...new Set([...new Set(addrs)])])]
```

Comment (dapplion) [reply src/connection-manager/dialer/index.ts:262]: - Good to see this removed in PR https://github.com/libp2p/js-libp2p/pull/1498/files#diff-f41a0efb6793f2f536fc7b7eee135829b88f2aea74c0391ddb79d382871ca54fL238

**Pull request merged**